Use weak layer references and loose matching for composer legend customisation

nyalldawson opened the pull request:
Allows legend customisation to be restored when loading a composer
template

Fix #2738

This is an updated fix based on the fix used for 2.18 (https://github.com/qgis/QGIS/pull/4054), which addresses partially the comments from https://github.com/qgis/QGIS-Enhancement-Proposals/issues/87#issuecomment-275611243

I plan on extending this approach to allow resolving references to map layers within composer maps which have been set to locked layers, to allow templates with locked layers to be restored in different projects correctly.

nyalldawson committed:
Use weak layer references and loose matching for composer legend
customisation

Allows legend customisation to be restored when loading a composer
template

Fix #2738

nyalldawson committed:
sipify QgsLayerTreeGroup

nyalldawson committed:
sipify QgsLayerTreeLayer

nyalldawson committed:
Add missing docs

wonder-sk reviewed (approved):
Looks good to me!

One suggestion - we could extract the duplicated code for retrieval of data provider name into a utility method - but it is not a big deal...

nyalldawson commented:
> One suggestion - we could extract the duplicated code for retrieval of data provider name into a utility method - but it is not a big deal...

Are you ok with me adding a "providerKey" getter to QgsMapLayer for this purpose? It seemed odd to me while writing this PR that there wasn't a straightforward way to get this without all the casting and switches...

wonder-sk commented:
> Are you ok with me adding a "providerKey" getter to QgsMapLayer for this purpose?

Yeah I think that would be good!

nyalldawson committed:
[sipify] Always add virtual keyword for overridden methods

sip often requires the virtual keyword to be present, or
it chokes on covariant return types in overridden methods

nyalldawson committed:
Allow QgsDataProvider access through QgsMapLayer

Avoids having to cast to QgsRasterLayer/QgsVectorLayer before
accessing the common QgsDataProvider members

nyalldawson commented:
> Are you ok with me adding a "providerKey" getter to QgsMapLayer for this purpose?

I've done this a different way, by adding a QgsDataProvider getter to QgsMapLayer so that the common properties can all be accessed without the need for casting.
